src/main/java/com/williamfiset/algorithms/sorting/QuickSort3.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * Quicksort implementation using Hoare partitioning
+ *
+ * <p>Run with:
+ *
+ * <p>$ ./gradlew run -Palgorithm=sorting.QuickSort3
+ *
+ * @author Atharva Thorve, aaathorve@gmail.com
+ */
+package com.williamfiset.algorithms.sorting;
+
+import java.util.Random;
+
+public class QuickSort3 implements InplaceSort {
+  private static Random random = new Random();
+
+  @Override
+  public void sort(int[] values) {
+    QuickSort3.quickSort3(values);
+  }
+
+  public static void quickSort3(int[] ar) {
+    if (ar == null) return;
+    QuickSort3.randomizedQuickSort(ar, 0, ar.length - 1);
+  }
+
+  // partiton array in such a way that all the elements whose value is equal to
+  // pivot are grouped together
+  private static int[] partition3(int[] a, int l, int r) {
+    int j, k;
+    if (r - l <= 1) {
+      if (a[r] < a[l]) {
+        int temp = a[l];
```

Comment:
Maybe a swap helper function would be nice? You're using this swap paradigm in multiple places.

https://github.com/williamfiset/Algorithms/blob/master/src/main/java/com/williamfiset/algorithms/sorting/QuickSort.java#L50